pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/customobject/DistinctTable.java

```diff
@@ -0,0 +1,297 @@
+  private final Set<Record> _uniqueRecords;
+  private final PriorityQueue<Record> _sortedRecords;
+  private final List<Record> _records;
+
+  /**
+   * Constructor of the main {@code DistinctTable} which can be used to add records and merge other
+   * {@code DistinctTable}s.
+   */
+  public DistinctTable(DataSchema dataSchema, @Nullable List<SelectionSort> orderBy, int limit) {
+    _dataSchema = dataSchema;
+    _limit = limit;
+
+    // TODO: see if 10k is the right max initial capacity to use
+    // NOTE: When LIMIT is smaller than or equal to the MAX_INITIAL_CAPACITY, no resize is required.
+    int initialCapacity = Math.min(limit, MAX_INITIAL_CAPACITY);
+    _uniqueRecords = new ObjectOpenHashSet<>(initialCapacity);
+    if (orderBy != null) {
+      String[] columns = dataSchema.getColumnNames();
+      int numColumns = columns.length;
+      Object2IntOpenHashMap<String> columnIndexMap = new Object2IntOpenHashMap<>(numColumns);
+      for (int i = 0; i < numColumns; i++) {
+        columnIndexMap.put(columns[i], i);
+      }
+      int numOrderByColumns = orderBy.size();
+      int[] orderByColumnIndexes = new int[numOrderByColumns];
+      boolean[] orderByAsc = new boolean[numOrderByColumns];
+      for (int i = 0; i < numOrderByColumns; i++) {
+        SelectionSort selectionSort = orderBy.get(i);
+        orderByColumnIndexes[i] = columnIndexMap.getInt(selectionSort.getColumn());
+        orderByAsc[i] = selectionSort.isIsAsc();
+      }
+      _sortedRecords = new PriorityQueue<>(initialCapacity, (record1, record2) -> {
+        Object[] values1 = record1.getValues();
+        Object[] values2 = record2.getValues();
+        for (int i = 0; i < numOrderByColumns; i++) {
+          Comparable valueToCompare1 = (Comparable) values1[orderByColumnIndexes[i]];
+          Comparable valueToCompare2 = (Comparable) values2[orderByColumnIndexes[i]];
+          int result =
+              orderByAsc[i] ? valueToCompare2.compareTo(valueToCompare1) : valueToCompare1.compareTo(valueToCompare2);
+          if (result != 0) {
+            return result;
+          }
+        }
+        return 0;
+      });
+    } else {
+      _sortedRecords = null;
+    }
+    _records = null;
+  }
+
+  /**
+   * Returns the {@code DataSchema} of the {@code DistinctTable}.
+   */
+  public DataSchema getDataSchema() {
+    return _dataSchema;
+  }
+
+  /**
+   * Returns the number of unique records within the {@code DistinctTable}.
+   */
+  public int size() {
+    if (_uniqueRecords != null) {
+      // Server-side
+      return _uniqueRecords.size();
+    } else {
+      // Broker-side
+      return _records.size();
+    }
+  }
+
+  /**
+   * Adds a record into the DistinctTable and returns whether more records should be added. No more records should be
+   * added iff there is no order-by column and enough unique records have been collected.
+   */
+  public boolean add(Record record) {
+    if (_uniqueRecords.contains(record)) {
```

Comment:
We should not be checking for contains() on every add(). This will hurt performance since the add() method on the hashset() later on will anyway do a contains internally and will accordingly decide to add or ignore the new item. It will also return true or false if the item was added or not. At least, the java hashset works like this. 

I don't see the purpose of this if-check in lines 126-128. There are two cases:

- **No order by:** simply do.add() and return if the limit has been reached or not
- **Order by:** If limit hasn't reached, add to both set and PQ. Else poll from PQ, delete from set and add to both. Always return true in this case.

Lines 129-145 handle both the cases.

Reply:
For non-order-by, yes we can save this extra contains check.

For order-by, we should still first check if it contains the record to avoid the unnecessary remove. More importantly, blindly adding record can result in unexpected expanding of the map.